NetLogo/Galapagos
# Events, queries, and auto-save

This PR adds a few features to NetLogo Web:

- An event system for model changes (authoring) and usage (widget interactions), that can be hooked into by users embedding models in other pages or by apps that are wrapping around NetLogo Web (like NetTango).
- A simple query system for getting information out of a model.  This is meant to help users embedding models as iframes who might want to extract data to use outside of NetLogo Web, perhaps in response to certain events.
- An auto-save feature built around the event system that allows resuming work in progress after page reloads and/or refreshes.  It's aware of the model source location, so it tracks progress on each URL, file upload, and New Model separately.  It also allows for quick reversion of changes made back to the original source.

### How to Test

For the event system, you can add the `debugEvents` query parameter to the URL of the `/launch` or `/web` pages to see the generated events logged to the web console.  If you're just reviewing the events, that should be enough.  For end-users embedding models as iframes by hot-linking or via static HTML exports, they also can use the `relayIframeEvents` query parameter on their model in order to get events automatically passed to their containing window for handling by code.  A full listing of events is contained in `listener-events.coffee`.

For the query system, you can add the `debugQueries` query parameter to the URL of the `/launch` page to attach a simple `makeQuery()` function to the `window`.  The query type strings for that method are listed in `debug-query-maker.coffee`.  The full query list is in `iframe-query-handler.coffee`.  In theory this query system doesn't just have to be used through an iframe.  Something like NetTango that embeds a NetLogo Web model could also use it, but at the moment I haven't made that switch.

The auto-save feature should be pretty self-explanatory (knock on wood).  I wound up not including a user prompt when they click to *Revert to Original*, because it wound up being pretty easy to add a *Undo Revert* button as well.  The ease of undoing the action if they didn't really want it made the prompt seem unnecessary.  I also did not add the *Set as Authoritative* button yet.  We can still look at that in the future.  I feel like it may be less useful if we have a real, granular undo/redo system in place.  But exporting a file or new model to an nlogo file will set the current authoritative source with the exported file name.

### Future Work

- Create documentation for the features, including tutorials for sending NetLogo Web event and model data off to remote storage for later review.  I'll be working on this over the next couple of weeks.
- Add the ability to raise user-defined events through the logging extension.
- Make sure NetTango is fully unified with the events system, and that all NetTango events are also exposed.
- Expand the auto-save feature into a true undo/redo system.  Ideally this would be a smarter system to track individual changes (like widget moves, plot renames, etc) so that they could be undone without a full session teardown and compile from nlogo source.
- Add the ability to hook into a user's remote storage, like GitHub, Dropbox, or Google Drive, in order to store work in progress there instead of just in the browser local cache.
- Add events for engine-specific things, where it makes sense.  At least prims like `reset-ticks`, `clear-all`, etc.  NetLogo desktop has events for turtle birth and death, which seems excessive to me, but I can see where having the option would be nice.
- Many, many more smaller items.
